app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt

```diff
@@ -115,37 +122,63 @@ class ConversationViewModel @Inject constructor(
     var establishedCallConversationId: ConversationId? = null
 
     init {
-        fetchMessages()
-        listenConversationDetails()
+        observeConversationDetailsAndMessages()
         fetchSelfUserTeam()
         setFileSharingStatus()
         listenOngoingCall()
         observeEstablishedCall()
     }
 
-    // region ------------------------------ Init Methods -------------------------------------
-    private fun fetchMessages() = viewModelScope.launch(dispatchers.io()) {
-        getMessageForConversation(conversationId).collect { messages ->
-            updateMessagesList(messages)
+    private fun observeConversationDetailsAndMessages() {
```

Comment:
IMO this looks a bit better, and no need to start with `null`

```
observeConversationDetails(conversationId)
    .onEach { conversationDetailsResult ->
        when (conversationDetailsResult) {
            is Failure -> handleConversationDetailsFailure(conversationDetailsResult.storageFailure)
            is Success -> handleConversationDetails(conversationDetailsResult.conversationDetails)
        }
    }
    .combine(getMessageForConversation(conversationId), ::Pair)
    .collect { (conversationDetailsResult: ObserveConversationDetailsUseCase.Result, uiMessages: List<UIMessage>) ->
        if (conversationDetailsResult is Success) {
            updateMessagesList(uiMessages)
            when (val details = conversationDetailsResult.conversationDetails) {
                is ConversationDetails.OneOne -> {
                    val lastUnreadMessage = details.lastUnreadMessage
                    if (lastUnreadMessage != null) {
                        uiMessages.firstOrNull { it.messageHeader.messageId == lastUnreadMessage.id }?.let {
                            conversationViewState = conversationViewState.copy(lastUnreadMessage = it)
                        }
                    }
                }
                is ConversationDetails.Group -> {
                    val lastUnreadMessage = details.lastUnreadMessage
                    if (lastUnreadMessage != null) {
                        uiMessages.firstOrNull { it.messageHeader.messageId == lastUnreadMessage.id }?.let {
                            conversationViewState = conversationViewState.copy(lastUnreadMessage = it)
                        }
                    }
                }
                else -> {}
            }
        }
    }
```

Reply:
looks nice indeed but I think combine() will not emit anything untill observeConversationDetails does not emit anything, so before we display messages we have to wait for a conversation details in my opinion it is a bit unnecessary performance decrease, let me know if I am wrong 🤔

Reply:
i think you're wrong...there is 

```
.onEach { conversationDetailsResult ->
    when (conversationDetailsResult) {
        is Failure -> handleConversationDetailsFailure(conversationDetailsResult.storageFailure) 
        is Success -> handleConversationDetails(conversationDetailsResult.conversationDetails)
    }
}
```
just before the combine, so it wont wait for `getMessageForConversation` to `handleConversationDetailsFailure` or `handleConversationDetails`